### ISSUE TITLE
itrim trimming where it shouldn't

### ISSUE THREAD

**Issue by luijs** [irods/irods#7468]
- [x] main
- [x] 4-3-stable
---
### iRODS Version, OS and Version
irods 4.3.1, ubuntu 20.04
Also seen in 4.3.0/ubuntu 20.04

### What did you try to do?
Trim a file that only had stale copies

### Expected behavior
The file would not be trimmed.
According to the docs here row 8: https://docs.irods.org/4.3.1/system_overview/data_objects/#trim
`8 | X | X | X | X | Data object has fewer than the minimum number of good replica`

If you have two stale replicas of a file and you do an itrim it should not delete any replica.

### Observed behavior (including steps to reproduce, if applicable)
The file was trimmed.

In more detail:
```
iput version.json
irepl -S hot_1 -R tape_1 version.json
iadmin modrepl logical_path /zone/home/user/version.json replica_number 1 DATA_REPL_STATUS 0
iadmin modrepl logical_path /zone/home/user/version.json replica_number 0 DATA_REPL_STATUS 0
ils -rl
/zone/home/user/:
  irods             0 hot_1          240 2024-01-19.14:25 X version.json
  irods             1 tape_1          240 2024-01-19.14:25 X version.json
  C- /zone/home/user/
/zone/home/user/:
itrim -N 1 version.json -S hot_1
Total size trimmed = 0.000 MB. Number of files trimmed = 1.
ils -rl
/zone/home/user/:
  irods             1 tape_1          240 2024-01-19.14:25 X version.json

```
I did have some rules in effect that were doing something in `pep_api_data_obj_trim_pre` and 'pep_api_data_obj_unlink_pre'. However, that should impact this, and if I uncommented those rules I still had the same behaviour. Curious to see if this is only my machine or if this can be replicated.

Note: When I put the replica_status of the file I wanted to keep to 2 or 3 the file was not trimmed as expected

**Comment by trel** [comment]
If you specify a replica (with -n or -S), then I think this is expected behavior.

That table you reference is what happens when there is no particular replica targeted.   We can make that more explicit in the documentation.

**Comment by luijs** [comment]
Hmm, is that really the expected behaviour? And is it also the desired behaviour? 
Why would this be different for repl_status 0 and repl_status 2, and why would it be different when you specify the exact replica to trim or not? Shouldn't blocking trimming be independent from if you specify a replica or not?

**Comment by luijs** [comment]
Some further testing shows that it does work when there is only 1 stale replica which makes it a bit of an edge case.

```
$ ils -rl
/zone/home/user/:
  irods             0 hot_1          240 2024-01-19.14:55 & version.json
  irods             1 tape_1          240 2024-01-19.14:55 X version.json
  C- /zone/home/user/
/zone/home/user/:
$ itrim -N 1 -S hot_1 version.json
Total size trimmed = 0.000 MB. Number of files trimmed = 0.
$ itrim -N 1 -S tape_1 version.json
Total size trimmed = 0.000 MB. Number of files trimmed = 1.
```

The situation when i don't add a resource to delete specifically gets weirder, because then all replica's are deleted:
```
$ ils -rl
/zone/home/user:
  irods             0 hot_1         240 2024-01-19.15:01 X version.json
  irods             1 tape_1          240 2024-01-19.15:01 X version.json
  C- /zone/home/user
/zone/home/user:
$ itrim -N 1 version.json
remote addresses: X.X.X.X ERROR: trimUtil: trim error for /zone/home/user/version.json.  status = -310000 USER_FILE_DOES_NOT_EXIST
Total size trimmed = 0.000 MB. Number of files trimmed = 0.
$ ils -rl
/zone/home/user:
```

**Comment by korydraughn** [comment]
After some discussion, we're thinking this needs to be changed such that `-N <minimum>` is always honored, even when all replicas are stale.

That means for case 8, one replica should be trimmed, leaving one good replica and one stale replica.

Until we settle on the appropriate behavior, the recommendation is to be explicit about which replica is to be trimmed. That effectively means always use `itrim -N <minimum> -n <replica_number> <data_object>`. The trim API will eventually require users to target a specific replica _(this approach can be seen in the HTTP API's interface)_.

**Comment by alanking** [comment]
So... some more observations...

Using `-n` to target a replica works fine:
```bash
$ ils -l foo
  alan              0 demoResc        10240 2024-01-19.17:31 X foo
  alan              1 third_resc        10240 2024-01-19.17:31 X foo
$ itrim -N1 -n0 foo
Total size trimmed = 0.010 MB. Number of files trimmed = 1.
$ ils -l foo
  alan              1 third_resc        10240 2024-01-19.17:31 X foo
```

And as it turns out, you don't even need `-N` with stales!
```bash
$ irepl foo
$ ils -l foo
  alan              1 third_resc        10240 2024-01-19.17:31 X foo
  alan              2 demoResc        10240 2024-01-19.17:31 X foo
$ itrim -n1 foo
Total size trimmed = 0.010 MB. Number of files trimmed = 1.
$ ils -l foo
  alan              2 demoResc        10240 2024-01-19.17:31 X foo
```

In fact, you don't need `-N` to reproduce the original issue, either!
```bash
$ itrim foo
remote addresses: 127.0.0.1 ERROR: trimUtil: trim error for /tempZone/home/public/foo.  status = -310000 USER_FILE_DOES_NOT_EXIST
Total size trimmed = 0.000 MB. Number of files trimmed = 0.
$ ils -l foo
remote addresses: 127.0.0.1 ERROR: lsUtil: srcPath /tempZone/home/public/foo does not exist or user lacks access permission
```
And the output lies - it was trimmed.

----

Anyway, back to `-N`:
```
 -N  numCopies - minimum number of most current copies to keep. The default
     value is 2.
```

This help text, while impressively imprecise, has historically been understood to mean "minimum number of good replicas to keep". That is how the API is implemented, and it is the way that it has been documented in the link shared above. Beyond that, there are rules governing the magic decision-making done by trim to determine which replicas to be rid of.

The expected behavior suggested by the historical interpretation of `-N` would be that none of the replicas are trimmed because there are no good replicas. This is what is documented in the table (even if it is not implemented correctly). For completeness, not providing a `-N` would mean "trim down to 2 good replicas".

The suggestion is to change the definition of `-N` to mean "minimum number of replicas to keep", discarding the "good" qualifier. 

So, let's make a more thorough table. I'm going to fill it with what I think should happen and maybe we can reverse engineer some rules from that. :) Assume that the replicas are ordered oldest to newest from left to right, replica numbers 0 through 3. 'X' is a stale replica, '&' is a good replica, and '-' means "doesn't exist" or "removed". Let's just say we run `itrim` on the object with no parameters.

| case | start | end |
| ---- | ---- | ---- |
| 0 | XXXX | --XX |
| 1 | XXX& | --X& |
| 2 | XX&X | --&X |
| 3 | XX&& | --&& |
| 4 | X&XX | -&-X |
| 5 | X&X& | -&-& |
| 6 | X&&X | -&&- |
| 7 | X&&& | --&& |
| 8 | &XXX | &--X |
| 9 | &XX& | &--& |
| 10 | &X&X | &-&- |
| 12 | &X&& | --&& |
| 13 | &&XX | &&-- |
| 14 | &&X& | -&-& |
| 15 | &&&X | -&&- |
| 16 | &&&& | --&& |

So... here's the priority order of replicas as I understand it based on the table above, if we're all in agreement that that is the expected behavior, from most likely to be trimmed to least likely to be trimmed. I guess it's more like a spectrum...
 - oldest stale replica
 - ...the other stale replicas...
 - newest stale replica
 - oldest good replica
 - ...the other good replicas...
 - newest good replica

The critical point is that there is a classification of stale versus good replicas, with stale replicas taking priority.

And then we trim down with some algorithm like this:
```
while (!stale_replica_list.empty() && total_replica_count > minimum_replica_count) {
    trim_oldest_replica(stale_replica_list);
    update_total_replica_count();
}

while (total_replica_count > minimum_replica_count) {
    trim_oldest_replica(good_replica_list);
    update_total_replica_count();
}
```

This ensures that the total replica count never falls below 2 and all of the stale replicas are trimmed before the good replicas are touched.

So, taking case 14, the logic might look like this...
```
&XX&
0123
```

stale_replica_list: 1,2
good_replica_list: 0,3

1 and 2 are trimmed and now the stale replica list is empty. The minimum replica count defaults to 2, so no good replicas are trimmed. At the end we are left with:
```
&--&
0123
```

Or case 15...
```
&&&X
0123
```

stale_replica_list: 3
good_replica_list: 0,1,2

Replica 3 is trimmed and now the stale replica list is empty. The minimum replica count defaults to 2 and we have 3 replicas left, so we trim the oldest good replica. Replica 0 is trimmed and now we have the minimum replica number, so we are done. At the end we are left with:
```
-&&-
0123
```

----

This would be a departure from how trim has worked in the past, but I don't think it would be too terribly different. Like @korydraughn said, being more precise with the trim API is always better because it takes a little bit of the guesswork out of what it's going to do.

**Comment by trel** [comment]
The 'impressively imprecise' language in `itrim -h` was cleaned up a couple weeks ago and will be included in 4.3.2...

https://github.com/irods/irods_client_icommands/commit/1df6256718a640368290738f17cf0f5b50d44e85

**Comment by trel** [comment]
And agreed, changing the algorithm to be more deterministic in the face of stale replicas is a good thing.

The question about replica_status of 2 or 3 is generally a separate case as those indicate the replica in question is locked.  I do not expect `trim()` to ever touch/affect those replicas (as they are 'in flight' and off limits or 'stuck' and already in need of admin attention).
